File: Books/Entendendo_Algoritmos/Capítulo 1/Pesquisa Binária/TypeScript/ts/pag_26_27.ts
// Função para realizar pesquisa binária
function pesquisaBinaria(lista: number[], item: number): number | null {
  // Define os limites inicial e final da pesquisa
  let baixo: number = 0;
  let alto: number = lista.length - 1;

  // Enquanto o limite inferior não ultrapassar o limite superior
  while (baixo <= alto) {
    // Calcula o índice do elemento central
    const meio: number = Math.floor((baixo + alto) / 2);
    const chute: number = lista[meio];

    // Verifica se o elemento central é o item procurado
    if (chute === item) {
      return meio; // Retorna o índice se encontrado
    }

    // Se o valor do elemento central for maior que o item
    if (chute > item) {
      alto = meio - 1; // Ajusta o limite superior para o meio - 1
    } else {
      baixo = meio + 1; // Ajusta o limite inferior para o meio + 1
    }
  }

  // Retorna null se o item não for encontrado
  return null;
}

// Lista para teste
const minhaLista: number[] = [1, 3, 5, 7, 9];

// Testa a função de pesquisa binária
console.log(pesquisaBinaria(minhaLista, 3)); // => 1 (índice do item 3 na lista)
console.log(pesquisaBinaria(minhaLista, -1)); // => null (item -1 não encontrado)

File: Books/Entendendo_Algoritmos/Capítulo 1/Pesquisa Binária/Python/pag_26_27.py
# baixo = 0
# alto = len(lista) - 1
# meio = (baixo + alto) // 2
# meio será arredondado para baixo automaticamente pelo Python se (baixo+alto) não for um número par. Se o chute for muito baixo, você atualizará a variável baixo proporcionalmente.

# if chute < item:
#     baixo = meio + 1

# Se o chute for muito alto, você atualizará a variável alto. Aqui está o código completo:


# baixo e alto acompanham a parte da lista que você está procurando.
def pesquisa_binaria(lista, item):
    baixo = 0
    alto = len(lista) - 1

    # Enquanto ainda não conseguiu chegar a um único elemento...
    while baixo <= alto:

        # ...verifica o elemento central.
        meio = (baixo + alto) // 2
        chute = lista[meio]

        # Acha o item.
        if chute == item:
            return meio

        # O chute foi muito alto.
        if chute > item:
            alto = meio - 1

            # O chute foi muito baixo.
        else:
            baixo = meio + 1

            # O item não existe.
    return None


# Vamos testá-lo!
minha_lista = [1, 3, 5, 7, 9]
# Lembre-se, as listas começam no 0. O próximo endereço tem índice 1.
print(pesquisa_binaria(minha_lista, 3))  # => 1
# "None" significa nulo em Python. Ele indica que o item não foi encontrado.
print(pesquisa_binaria(minha_lista, -1))  # => None
